File: bindings/swift/Sources/Proven/SafeML.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Machine learning utilities delegated to libproven FFI.
///
/// Activation functions (softmax, sigmoid, ReLU) via the
/// formally verified Idris 2 core.

import CProven

/// Machine learning utilities backed by libproven.
public enum SafeML {
    /// Softmax activation function over an array of values.
    /// Returns the softmax-normalized output, or an error.
    public static func softmax(_ input: [Double]) -> Result<[Double], ProvenError> {
        guard !input.isEmpty else { return .success([]) }

        var output = [Double](repeating: 0.0, count: input.count)
        let status = input.withUnsafeBufferPointer { inBuf in
            output.withUnsafeMutableBufferPointer { outBuf in
                proven_ml_softmax(inBuf.baseAddress, outBuf.baseAddress, input.count)
            }
        }
        if let error = ProvenError.fromStatus(status) {
            return .failure(error)
        }
        return .success(output)
    }

    /// Sigmoid activation: 1 / (1 + exp(-x)).
    public static func sigmoid(_ x: Double) -> Double {
        proven_ml_sigmoid(x)
    }

    /// ReLU activation: max(0, x).
    public static func relu(_ x: Double) -> Double {
        proven_ml_relu(x)
    }
}

File: bindings/php/src/Proven/SafeML.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeML - FFI wrapper for proven_ml_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe machine learning activation functions via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 */
final class SafeML
{
    /**
     * Compute softmax over an array of values.
     *
     * @param float[] $input Array of input values.
     * @return float[]|null Array of softmax probabilities, or null on error.
     */
    public static function softmax(array $input): ?array
    {
        $len = count($input);
        if ($len === 0) {
            return null;
        }
        $inBuf = \FFI::new("double[$len]");
        $outBuf = \FFI::new("double[$len]");
        for ($i = 0; $i < $len; $i++) {
            $inBuf[$i] = (float) $input[$i];
        }
        $status = FFI::getLib()->proven_ml_softmax(
            \FFI::addr($inBuf[0]), \FFI::addr($outBuf[0]), $len
        );
        if ($status !== 0) {
            return null;
        }
        $output = [];
        for ($i = 0; $i < $len; $i++) {
            $output[] = (float) $outBuf[$i];
        }
        return $output;
    }

    /**
     * Sigmoid activation function: 1 / (1 + exp(-x)).
     */
    public static function sigmoid(float $x): float
    {
        return FFI::getLib()->proven_ml_sigmoid($x);
    }

    /**
     * ReLU activation function: max(0, x).
     */
    public static function relu(float $x): float
    {
        return FFI::getLib()->proven_ml_relu($x);
    }

    /**
     * Leaky ReLU activation function: x if x > 0, alpha * x otherwise.
     */
    public static function leakyRelu(float $x, float $alpha): float
    {
        return FFI::getLib()->proven_ml_leaky_relu($x, $alpha);
    }

    /**
     * Clamp a value to a range.
     */
    public static function clamp(float $x, float $minVal, float $maxVal): float
    {
        return FFI::getLib()->proven_ml_clamp($x, $minVal, $maxVal);
    }
}
